### PR TITLE
ROM Train on DataSet

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Addresses #731 for ROMs (not PP).

##### What are the significant changes in functionality due to this change request?
Allows option to pass training data sets directly to ROM SupervisedLearning algorithms rather than converting everything to dictionaries.

A flag is used to allow the SVL to self-identify whether it needs legacy training (dictionaries) or can handle training via DataSet.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/Models/ROM.py [hunk 1]
```diff
@@ -307,64 +308,91 @@ def train(self,trainingSet):
...
       # TODO: The following check may need to be moved to Dummy Class -- wangc 7/30/2018
-      if type(trainingSet).__name__ != 'dict' and trainingSet.type == 'HistorySet':
+      if type(trainingSet) != dict and trainingSet.type == 'HistorySet':
         if not trainingSet.checkIndexAlignment(indexesToCheck=self.pivotParameterId):
           self.raiseAnError(IOError, "The data provided by the data object", trainingSet.name, "is not synchonized!",
                   "The time-dependent ROM requires all the histories are synchonized!")
-      self.trainingSet = copy.copy(self._inputToInternal(trainingSet))
+      # LEGACY SupervisedLearning (SVL) objects train on dictionaries/matrices
+      # New SVL can bypass the data manip and use the dataset directly
+      useDict = self.supervisedContainer[0].needsDictTraining
```
Comment (PaulTalbot-INL) [inline framework/Models/ROM.py:322]: Here is the use of the check for legacy training (needsDictTraining)

File: framework/SupervisedLearning/SupervisedLearning.py [hunk 2]
```diff
@@ -131,6 +131,7 @@ def __init__(self):
     self.muAndSigmaFeatures = {}   # normalization parameters
     self.metadataKeys = set()      # keys that can be passed to DataObject as meta information
     self.metadataParams = {}       # indexMap for metadataKeys to pass to a DataObject as meta dimensionality
+    self.needsDictTraining = True  # True if the "train" method expects a dictionary ONLY
```
Comment (PaulTalbot-INL) [inline framework/SupervisedLearning/SupervisedLearning.py:134]: Here's the defaulting declaration of the legacy node flag. Existing SVLs will all use dicts to train, but this should be discouraged for future algorithms.

**Comment by aalfonsi** [comment]
@PaulTalbot-INL @wangcj05 @mandd guys, is this going to happen anytime soon?

**Comment by PaulTalbot-INL** [comment]
>  guys, is this going to happen anytime soon?

If I recall correctly, the DMDc merge introduced some conflicting mechanics as well as code conflicts, so I haven't returned to try and finish it off. If you need it for something, I can try and fit it in (you know, late evening or early mornings)

**Comment by aalfonsi** [comment]
> > guys, is this going to happen anytime soon?
> 
> If I recall correctly, the DMDc merge introduced some conflicting mechanics as well as code conflicts, so I haven't returned to try and finish it off. If you need it for something, I can try and fit it in (you know, late evening or early mornings)

oh got it. Do you have an ETA more and less?

**Comment by PaulTalbot-INL** [comment]
You know how it is, as soon as someone needs it, it will become a priority. I don't foresee needing it until near end of this FY, for my stuff, if not early next year.

**Comment by aalfonsi** [comment]
> You know how it is, as soon as someone needs it, it will become a priority. I don't foresee needing it until near end of this FY, for my stuff, if not early next year.
Okay. Slow peace at the lab it seems. eheheh I am kidding.
Thanks for the update ;)

**Comment by aalfonsi** [comment]
@PaulTalbot-INL news on this?

**Comment by PaulTalbot-INL** [comment]
I believe all of the critical changes have been ported to other PRs and merged. @dylanjm can you confirm? I can close this if there's nothing else of value to scavenge from it.

**Comment by dylanjm** [comment]
@PaulTalbot-INL Yes, I was able to grab the custom sampler stuff in another PR. I think Yeni has a few changes in her RAVEN which she'll push as a PR eventually. Unless there are other changes, I think the majority of what we needed here has been integrated.

**Comment by PaulTalbot-INL** [comment]
Okay, I'll close this then. Thanks, @dylanjm and @aalfonsi!

**Pull request closed without merging**